### ISSUE TITLE
Login-PowerBI -Credential $cred does not work in Azure Automation

### ISSUE THREAD

**Issue by furmangg** [microsoft/powerbi-powershell#190]
This code doesn't work in Azure Automation:
`Login-PowerBI -Credential $mycreds`

The reason it doesn't work is that it's calling WindowsAuthenticationFactory.InitializeCache which tries to start a process. 
https://github.com/microsoft/powerbi-powershell/blob/master/src/Common/Common.Authentication/WindowsAuthenticationFactory.cs

That may make sense for interactive login popups, but for passing `-Credential` it shouldn't be interactive. (I don't think -Credential currently supports multi-factor auth logins.)

To prove launching a process is the culprit, the following code in a Azure Automation Runbook (PowerShell) will cause it to fail and be suspended after 3 retries just like the `Login-PowerBI -Credential $cred` code does:

````
$ps = new-object System.Diagnostics.Process
$ps.StartInfo.Filename = "ipconfig.exe"
$ps.StartInfo.Arguments = " /all"
$ps.StartInfo.RedirectStandardOutput = $True
$ps.StartInfo.UseShellExecute = $false
$ps.start() | Out-Null
$ps.WaitForExit()
[string] $Out = $ps.StandardOutput.ReadToEnd();
$Out
````

To also prove you can't run an executable inside Azure Automation, see this link:

https://feedback.azure.com/forums/246290-automation/suggestions/31957750-can-we-execute-exe-or-bat-files-inside-runbook

The reason I'm trying to do credential auth is because service principal auth doesn't currently have access to classic workspaces and to administrative APIs I believe.

Please enhance the Login-PowerBI cmdlet to avoid launching a process when you use the `-Credential` parameter only.

**Comment by furmangg** [comment]
Maybe instead of executing the AzureADWindowsAuthenticator.exe you can simply run a few lines of PowerShell when the -Credential is specified (meaning non-interactive login):

```
$body=@{
    "grant_type"="password";
    "resource"="$resourceAppIdURI";
    "client_id"=$environment.AzureADClientId;
    "username"=$username;
    "password" = $password;
}

$authority = "https://login.windows.net/$tenant/oauth2/token";
 
$result=Invoke-RestMethod -Uri $authority -Method POST -Body $body 
$result.access_token
```

**Comment by RenjuRNair** [comment]
@furmangg 
 Are you able to do **Login-PowerBI -Credential $cred** from runbook?
Can you please explain more on the solution you provided above?

**Comment by danieljohansso** [comment]
I have also been trying this and have not been able to solve it or find a workaround. Is there any update regarding this. 

I'm running an Azure Runbook with the following code:

```
    $Credentials= Get-AutomationPSCredential -Name 'Credentials'
    
    Connect-PowerBIServiceAccount -Credential $Credentials
```
Is there any known workarounds? This feature is critical! Running Powershell local is not a solution..

**Comment by melmsater** [comment]
Any updates on this issue? 
Running into the same problem in Azure Runbook.

**Comment by bee911** [comment]
any update as still hangs in azure runbook at Connect-PowerBIServiceAccount -Credential so i am unable to automate pbi!

**Comment by alliedcvil** [comment]
Any update on this issue? We can't do much of anything as far as automation without this functionality since we utilize Azure Automation.

**Comment by bee911** [comment]
> Any update on this issue? We can't do much of anything as far as automation without this functionality since we utilize Azure Automation.

since msft silent as usual, i opened support ticket with azure few weeks which of course unresolved.  they escalated to engineering and came back with below response:  lol

"We suggest you use a hybrid runbook worker to execute your script, because the cmdlets with troubles,  are requiring elevated access and aren't supported to run in cloud job to provide necessary security & isolation to the SQLPAL instance allocated for a job to run. " 

this is crazy as i am not standing up an azure vm or on-prem vm to run to run automation hybrid.  i agree would work perfectly since running win os but that is way to costly

even though case still open, i am not getting getting anywhere and needed solution so went in totally different direction...what i did is created an azure function (consumption based) with version of 2, set execution to be powershell, set function to import powerbi modules and then created timer function.  i ported code to azure function and can attest it works perfectly.  so i'd recommend doing that as clearly will not work in automation.

**Comment by yuanman2016** [comment]
I searched a long time of how to bypass interactive login to PowerBI with a user but no lucky. service principle is an option but need to enable in Power BI tenant level which doesn't work for me as my company doesn't enable it.  Even token works, but to get an AAD token also involve interactive login with PowerBI service.  The way I am having now is running a front PowerShell script with while(1) to keep processing PowerBI tasks such as refresh dataset within interval you set

**Comment by furmangg** [comment]
Is there a cmdlet which takes a token?

**Comment by yuanman2016** [comment]
There are some powerbi modules have command to get  PowerBI auth token, such as POWERBIPS which has a command call Get-PBIAuthToken. Or you can use Connect-PowerBIServiceAccount which belongs to the module: MicrosoftPowerBIMgmt.Profile. After you run the connect command, you actually don't need a token then can do some PowerBI tasks such as get workspace, dataset etc.

**Comment by furmangg** [comment]
@yuanman2016 I was just responding to your comment “Even token works, but to get an AAD token also involve interactive login with PowerBI service.” and wondering if we were able to get a login token in a non-interactive manner if that would help. I wasn’t aware of a cmdlet to login by passing in a token. And I think you have confirmed one doesn’t exist. Thanks.

**Comment by fredrikheden** [comment]
I'm facing this issue too, even when trying connect using ServicePrincipal:
`Connect-PowerBIServiceAccount -ServicePrincipal -ApplicationId $connection.ApplicationID -CertificateThumbprint $connection.CertificateThumbprint`
Any plans fixing this?

**Comment by furmangg** [comment]
Don’t you have to include the -Tenant switch when you use a service principal?
https://docs.microsoft.com/en-us/powershell/module/microsoftpowerbimgmt.profile/Connect-PowerBIServiceAccount?view=powerbi-ps

Off the top of my head you could probably rather put your login email suffix (like joe@parts.com would put parts.com) or you could lookup your tenant ID here:
https://www.whatismytenantid.com/

If that’s not it share your error message.

**Comment by fredrikheden** [comment]
@furmangg You are absolutely right. When the tenant is added it works with a service principal inside a runbook. Thanks!

**Comment by kartikjindgar** [comment]
@fredrikheden Can you please share the piece of code you used to login using service principal. I have been trying to login using Client credentials and Tenant ID but have not been able to make it work

**Comment by furmangg** [comment]
@kartikjindgar i assume that you gave the service principal access to the workspace and you also changed the Power BI admin setting allowing this (or allowing all) service principals to connect?

If you are using a secret to authenticate then:

Connect-PowerBIServiceAccount        -Credential <PSCredential>
       [-ServicePrincipal]
       [-Tenant <String>]

Here is a more complete example:
https://github.com/microsoft/PowerBI-Tools-For-Capacities/issues/14#issuecomment-652384543

If you are using a certificate then post back.

**Comment by kartikjindgar** [comment]
@furmangg thanks for such a prompt response. I was running the following command -

**Connect-PowerBIServiceAccount -Credential MyPSCredential -ServicePrincipal -Tenant yourTenantID**

It runs fine on my local machine but when I try to run it through a gitlab cicd pipeline it fails. I am getting the following error-

**Connect-PowerBIServiceAccount : Failed to populate environments in settings**
**+ Connect-PowerBIServiceAccount -ServicePrincipal -Credential $creds -T ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (Microsoft.Power...IServiceAccount:ConnectPowerBIServiceAccount) [Connect-Po 
   werBIServiceAccount], Exception
    + FullyQualifiedErrorId : Failed to populate environments in settings,Microsoft.PowerBI.Commands.Profile.ConnectPo 
   werBIServiceAccount**

Any clues how to solve this

**Comment by fredrikheden** [comment]
@kartikjindgar I'm using the following two lines to authenticate from a runbook:
```
$connection = Get-AutomationConnection -Name AzureRunAsConnection
Connect-PowerBIServiceAccount -Tenant "4********************5" -ServicePrincipal -ApplicationId $connection.ApplicationID -CertificateThumbprint $connection.CertificateThumbprint
```

**Comment by Mimiix** [comment]
Is there any progress on this issue? I can't use a serviceprincipal as the API that i use (https://docs.microsoft.com/en-us/rest/api/power-bi/imports/postimportingroup) doesn't support it.

**Comment by khouloudbelhaj** [comment]
Any update on this issue Please?

**Comment by yuanman2016** [comment]
> Any update on this issue?

Just received this update, want to share my findings:

I recently just use below way can by passing MFA:

$Cred = New-Object –TypeName System.Management.Automation.PSCredential –ArgumentList $user, $pw
Connect-PowerBIServiceAccount -Environment Public -Credential $Cred 

I am using a service account which is an email enabled account in AAD. After connecting successfully, I can do all APIs work which PowerBI provided

**Comment by furmangg** [comment]
@yuanman2016 where are you running that from? From Azure Automation Runbooks?

**Comment by yuanman2016** [comment]
Powershell and azure devops

Greg Galloway ***@***.***> 于 2021年5月17日周一 下午6:48写道：

**Comment by khouloudbelhaj** [comment]
@yuanman2016   Can I have more details please about the method you are referring to?I 
tried to execute the code you posted in a runbook but it still doesn't work for me.

**Comment by khouloudbelhaj** [comment]
@fredrikheden : I tried to work with  the code you proposed to resolve the problem but It's  not possible for my case to execute this part of the code : $connection = Get-AutomationConnection -Name AzureRunAsConnection , Is there any alternative?
![Capturerunasaccount](https://user-images.githubusercontent.com/47226862/118621324-8b18d480-b7c6-11eb-8155-c83d471e4c28.PNG)

**Comment by yuanman2016** [comment]
@khouloudbelhaj , I am using a service account, not a real user account which maybe a bit different. And I am running in local powershell, and Azure DevOps Azure Cli activity. I haven't tried Runbook yet.

**Comment by khouloudbelhaj** [comment]
@yuanman2016 : Do you have more details please about Azure Develops Azure Cli activity? I have tried to work with runbook but it failed to  connect to power BI. I switched to Azure functions and the AAD approach . The connexion  work locally via service principal but failed when it comes to function.

**Comment by CattieCat** [comment]
See if it works for you now since we released a new version 1.2.0 with MSAL.

**Comment by khouloudbelhaj** [comment]
@CattieCat Yes it worked for me , thank you for the job you have done! This will update will solve many problems.

**Comment by rangnar14** [comment]
> $Cred = New-Object –TypeName System.Management.Automation.PSCredential –ArgumentList $user, $pw

@yuanman2016 I am trying to get the powershell module for powerbi working through azure devops pipeline. 
Things are working fine, when I do it in powershell commandline. But when it goes via the azure devops ppeline I see the error which is mentioned in this thread.

You have mentioned that you are using Azure devops and Powershell and were able to get this accomplished right? 
Can you please provide the pipeline code part which does this. Mine looks like this:
$SecurePassword = ConvertTo-SecureString $PbiPassword -Force -AsPlainText
$PbiCredential = New-Object –TypeName System.Management.Automation.PSCredential –ArgumentList $PbiUser, $SecurePassword
Connect-PowerBIServiceAccount -Credential $PbiCredential -Environment Public 

Error is:
+ Connect-PowerBIServiceAccount -Credential $PbiCredential -Environment ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : WriteError: (Microsoft.Power...IServiceAccount:ConnectPowerBIServiceAccount) [Connect-Po 
   werBIServiceAccount], AggregateException
    + FullyQualifiedErrorId : One or more errors occurred.,Microsoft.PowerBI.Commands.Profile.ConnectPowerBIServiceAcc 
   ount

Request your help with this.

**Comment by yuanman2016** [comment]
@naraghavan , here is my code worked in 2021 when I posted, I didn't use it for a long time so not sure if it is still working, please test yourself:

param
(
    [parameter(Mandatory = $false)] [String] $pbixPath,
    [parameter(Mandatory = $false)] [String] $groupId,
    [parameter(Mandatory = $false)] [String] $datasetName
    )

#install module
Install-Module MicrosoftPowerBIMgmt.Profile -force -Scope CurrentUser

Import-Module  MicrosoftPowerBIMgmt.Profile 

Install-Module PowerBIPS -force -Scope CurrentUser

Import-Module  PowerBIPS 

$user = "user"
$pw = pwd 
$Cred = New-Object –TypeName System.Management.Automation.PSCredential –ArgumentList $user, $pw
Connect-PowerBIServiceAccount -Environment Public -Credential $Cred 

$token = Get-PowerBIAccessToken

$strToken = ($token.Values -join " ").Replace("Bearer ","").trim()

Get-PBIWorkspace -authToken $strToken

$dataset = Import-PBIFile -authToken $strToken -file $pbixPath.Replace("/","\") -dataSetName $datasetName -groupId $groupId -nameConflict CreateOrOverwrite


$dataset

#refresh data source credential
#$dataset = Get-PBIDataSet -authToken $strToken -name $datasetName -groupId $groupId

#Get-PBIDatasources -dataset $dataset.id -groupId $groupId -authToken $strToken
